sklearn/preprocessing/_polynomial.py

```diff
@@ -1008,16 +1062,37 @@ def transform(self, X):
                 for j in range(degree):
                     mask = X[:, i] < xmin
                     if np.any(mask):
-                        XBS[mask, i * n_splines + j] = (
-                            f_min[j] + (X[mask, i] - xmin) * fp_min[j]
-                        )
+                        linear_extr = f_min[j] + (X[mask, i] - xmin) * fp_min[j]
+                        if use_sparse:
+                            # Note: See comment about SparseEfficiencyWarning above.
+                            if not sparse.isspmatrix_lil(XBS_sparse):
+                                XBS_sparse = XBS_sparse.tolil()
+                            XBS_sparse[mask, j] = linear_extr
+                        else:
+                            XBS[mask, i * n_splines + j] = linear_extr
```

Comment:
The current support for using sparse matrices and for `sparse_output` add in several places several imbrications of `for`-loop and `if` in cascade which might be hard to read and maintain (the inner-most `if` here is at a 6th level of indentations for instance).

Do you think it is worth having two private methods respectively encapsulating the logic for working on CSR/LIL matrices and for working on dense arrays at the cost of some duplication?

Reply:
I see your point. The number of conditionals would not decrease. I'll have a look and try making it a bit more readable.

Reply:
I don't find any immediate simplification.